src/storage/tracked.rs

```diff
@@ -0,0 +1,378 @@
+use std::iter::Enumerate;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use std::slice;
+
+use hibitset::{BitSet, BitSetLike};
+
+use {Component, DenseVecStorage, Index, MaskedStorage, Storage, UnprotectedStorage};
+
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub enum Change {
+    #[doc(hidden)]
+    None,
+    Inserted,
+    Modified,
+    Removed,
+}
+
+impl Change {
+    fn add(&mut self, change: Change) {
+        let new = match (*self, change) {
+            (Change::None, new) => new,
+            (Change::Inserted, Change::Modified) => Change::Inserted,
+            (Change::Inserted, Change::Removed) => Change::None,
+            (Change::Modified, Change::Modified) => Change::Modified,
+            (Change::Modified, Change::Removed) => Change::Removed,
+            (Change::Removed, Change::Inserted) => Change::Modified,
+            (old, new) => panic!("Didn't expect change from {:?} to {:?}", old, new),
+        };
+
+        *self = new;
+    }
+}
+
+pub struct ChangeEvents<'a> {
+    inner: ChangeEventsInner<'a>,
+}
+
+impl<'a> Iterator for ChangeEvents<'a> {
+    type Item = (Index, Change);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner
+            .find(|&(_, &c)| c != Change::None)
```

Comment:
Why doesn't this use the bitset?

Reply:
The bit set is broken, it has to be removed.

Reply:
Can you elaborate bit why the bitset was broken? For me the only problem was that it wasn't being used in everywhere.

Reply:
It would have contained every component, no matter if it has changed or not. Thus, it was just an additional overhead and wasted memory.

Reply:
Ah, so you were adding to it even the `None` changes?

Reply:
The problem is that `get_mut` often gets called without an actual change. So I could only mark all eventually changed components by marking all components retrieved by `get_mut`, which in the worst case is then equal to MaskedStorage::bitset`.